File: rust/src/hold.rs
use crate::amount::parse_amount_minor;
use crate::transaction::{CanonicalPosting, validate_balanced_transaction};
use crate::{ValueError, ValueResult, required};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum HoldState {
    Open,
    Released,
    Settled,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ValueHold {
    pub hold_id: String,
    pub account_id: String,
    pub asset: String,
    pub amount_minor: String,
    pub state: HoldState,
}

fn validate_hold(hold: &ValueHold) -> ValueResult<i64> {
    required(&hold.hold_id, "Hold identity is required")?;
    required(&hold.account_id, "Hold account is required")?;
    required(&hold.asset, "Hold asset is required")?;
    let amount = parse_amount_minor(&hold.amount_minor)?;
    if amount <= 0 {
        return Err(ValueError::new("Hold amount must be positive"));
    }
    Ok(amount)
}

/// # Errors
/// Returns [`ValueError`] when a hold is malformed or exceeds available value.
pub fn create_value_hold(
    hold_id: &str,
    account_id: &str,
    asset: &str,
    amount_minor: &str,
    available_balance_minor: &str,
) -> ValueResult<ValueHold> {
    required(hold_id, "Hold identity is required")?;
    required(account_id, "Hold account is required")?;
    required(asset, "Hold asset is required")?;
    let amount = parse_amount_minor(amount_minor)?;
    let available = parse_amount_minor(available_balance_minor)?;
    if amount <= 0 {
        return Err(ValueError::new("Hold amount must be positive"));
    }
    if available < amount {
        return Err(ValueError::new("Insufficient available value for hold"));
    }
    Ok(ValueHold {
        hold_id: hold_id.into(),
        account_id: account_id.into(),
        asset: asset.into(),
        amount_minor: amount.to_string(),
        state: HoldState::Open,
    })
}

/// # Errors
/// Returns [`ValueError`] unless the hold is open.
pub fn release_value_hold(hold: &ValueHold) -> ValueResult<ValueHold> {
    if hold.state != HoldState::Open {
        return Err(ValueError::new("Only an open hold can be released"));
    }
    validate_hold(hold)?;
    let mut released = hold.clone();
    released.state = HoldState::Released;
    Ok(released)
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct HoldSettlement {
    pub hold: ValueHold,
    pub settled_amount_minor: String,
    pub released_amount_minor: String,
    pub postings: Vec<CanonicalPosting>,
}

/// # Errors
/// Returns [`ValueError`] unless an open hold can settle to a distinct destination.
pub fn settle_value_hold(
    hold: &ValueHold,
    destination_account_id: &str,
    amount_minor: Option<&str>,
) -> ValueResult<HoldSettlement> {
    if hold.state != HoldState::Open {
        return Err(ValueError::new("Only an open hold can be settled"));
    }
    let held_amount = validate_hold(hold)?;
    required(
        destination_account_id,
        "Hold settlement destination account is required",
    )?;
    if destination_account_id == hold.account_id {
        return Err(ValueError::new(
            "Hold settlement requires a distinct destination account",
        ));
    }
    let settled = parse_amount_minor(amount_minor.unwrap_or(&hold.amount_minor))?;
    if settled <= 0 || settled > held_amount {
        return Err(ValueError::new(
            "Hold settlement amount must be positive and cannot exceed the hold",
        ));
    }
    let postings = vec![
        CanonicalPosting {
            account_id: hold.account_id.clone(),
            asset: hold.asset.clone(),
            amount_minor: (-settled).to_string(),
        },
        CanonicalPosting {
            account_id: destination_account_id.into(),
            asset: hold.asset.clone(),
            amount_minor: settled.to_string(),
        },
    ];
    validate_balanced_transaction(&postings)?;
    let mut settled_hold = hold.clone();
    settled_hold.state = HoldState::Settled;
    Ok(HoldSettlement {
        hold: settled_hold,
        settled_amount_minor: settled.to_string(),
        released_amount_minor: (held_amount - settled).to_string(),
        postings,
    })
}

File: typescript/src/hold.ts
import { parseAmountMinor } from "./amount.js";
import {
  validateBalancedTransaction,
  type CanonicalPostingInput,
} from "./transaction.js";

export type HoldState = "open" | "released" | "settled";

export interface ValueHold {
  readonly holdId: string;
  readonly accountId: string;
  readonly asset: string;
  readonly amountMinor: string;
  readonly state: HoldState;
}

export interface HoldSettlement {
  readonly hold: ValueHold;
  readonly settledAmountMinor: string;
  readonly releasedAmountMinor: string;
  readonly postings: readonly CanonicalPostingInput[];
}

function requireIdentity(value: string, field: string): void {
  if (value.trim().length === 0) throw new Error(`${field} is required`);
}

function validateHold(hold: ValueHold): bigint {
  requireIdentity(hold.holdId, "Hold identity");
  requireIdentity(hold.accountId, "Hold account");
  requireIdentity(hold.asset, "Hold asset");
  const amount = parseAmountMinor(hold.amountMinor);
  if (amount <= 0n) throw new Error("Hold amount must be positive");
  return amount;
}

export function createValueHold(input: {
  readonly holdId: string;
  readonly accountId: string;
  readonly asset: string;
  readonly amountMinor: string;
  readonly availableBalanceMinor: string;
}): ValueHold {
  requireIdentity(input.holdId, "Hold identity");
  requireIdentity(input.accountId, "Hold account");
  requireIdentity(input.asset, "Hold asset");
  const amount = parseAmountMinor(input.amountMinor);
  const available = parseAmountMinor(input.availableBalanceMinor);
  if (amount <= 0n) throw new Error("Hold amount must be positive");
  if (available < amount)
    throw new Error("Insufficient available value for hold");
  return Object.freeze({
    holdId: input.holdId,
    accountId: input.accountId,
    asset: input.asset,
    amountMinor: amount.toString(),
    state: "open",
  });
}

export function releaseValueHold(hold: ValueHold): ValueHold {
  if (hold.state !== "open")
    throw new Error("Only an open hold can be released");
  validateHold(hold);
  return Object.freeze({ ...hold, state: "released" });
}

export function settleValueHold(input: {
  readonly hold: ValueHold;
  readonly destinationAccountId: string;
  readonly amountMinor?: string;
}): HoldSettlement {
  if (input.hold.state !== "open")
    throw new Error("Only an open hold can be settled");
  const held = validateHold(input.hold);
  requireIdentity(
    input.destinationAccountId,
    "Hold settlement destination account",
  );
  if (input.destinationAccountId === input.hold.accountId) {
    throw new Error("Hold settlement requires a distinct destination account");
  }
  const settled = parseAmountMinor(input.amountMinor ?? input.hold.amountMinor);
  if (settled <= 0n || settled > held) {
    throw new Error(
      "Hold settlement amount must be positive and cannot exceed the hold",
    );
  }
  const postings = Object.freeze([
    Object.freeze({
      accountId: input.hold.accountId,
      asset: input.hold.asset,
      amountMinor: (-settled).toString(),
    }),
    Object.freeze({
      accountId: input.destinationAccountId,
      asset: input.hold.asset,
      amountMinor: settled.toString(),
    }),
  ]);
  validateBalancedTransaction(postings);
  return Object.freeze({
    hold: Object.freeze({ ...input.hold, state: "settled" }),
    settledAmountMinor: settled.toString(),
    releasedAmountMinor: (held - settled).toString(),
    postings,
  });
}
